Delete remote branch

mdonatas opened the pull request:
Introduces a new form to delete remote branches from a FormBrowse commit context menu.
This PR is suggested solution for issue #1583

(sorry for diff noise, there is a separate commit dealing with some style/formatting issues)

Changes proposed in this pull request:
 - List remote branches in a commit context menu under local branches.
 
Screenshots:
This is I guess the most complex demo, deleting several (3) remote branches from two different remotes.
![test2](https://user-images.githubusercontent.com/483659/32701164-7e898670-c7c8-11e7-8866-80f8289354f3.gif)

Considerations:
- Maybe we don't need support for multiple remotes at the same time? That would simplify code a little.

RussKie commented:
Very nice.
I would probably change the signle-line dropdown with a checkbox list (which can grow or shrink with the form). this way there is no need for the second dialog and makes it obvious to the user that multi selection is possible

RussKie commented:
Funnily enough I was making FormBrowse Resharper green last night too.
Could you please leave "Cleanup some warnings and fix style inconsistencies" commit out. We can push it separately.

RussKie commented on GitUI/CommandsDialogs/FormDeleteRemoteBranch.cs:
```diff
@@ -0,0 +1,119 @@
+
+        private void DeleteRemote_CheckedChanged(object sender, EventArgs e)
+        {
+            Delete.Enabled = ((CheckBox)sender).Checked;
```
please change it to `Delete.Enabled = DeleteRemote.Checked;` this way there is no unsafe cast.

mdonatas commented:
I removed formatting/style changes from unrelated files and fixed the unsafe cast in the event handler.

As for checkboxes I am not too sure because:
- I bet 90%+ of the time no one will ever open that drop-down as it is pre-filled with the selection a user has already made in the menu
- It makes UI cleaner regardless of how many branches are on a commit
- Secondary dialog gives a list of all remote branches, not just the ones on a commit, since this seems like a power-user workflow it's better to keep it out of sight
- UI similarity with local branch delete
- And the real reason... (only half jokingly) I am lazy :)

XopherThomas commented:
I love it.  This feature would completely satisfy my request.  I think it adds a significantly safe amount of confirmation to avoid inadvertently deleting a remote branch.

RussKie commented:
I have envisaged something like this:

![image](https://user-images.githubusercontent.com/4403806/32766643-faf98a42-c963-11e7-87c5-b40011d7407c.png)

I'd think it be simpler to implement and use.... Thoughts?

vbjay commented:
Yes. I think adding a remote drop-down would be helpful too.

On Tue, Nov 14, 2017, 1:48 AM RussKie <notifications@github.com> wrote:

> I have envisaged something like this:
>
> [image: image]
> <https://user-images.githubusercontent.com/4403806/32766643-faf98a42-c963-11e7-87c5-b40011d7407c.png>
>
> I'd think it be simpler to implement and use.... Thoughts?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4146#issuecomment-344162389>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsV509zOvel3t1ocE91sos2alJkqZks5s2TfagaJpZM4Qa9Bv>
> .
>

mdonatas commented:
For the vast majority of users and their use cases I see this to be the workflow:
_You see a branch you don't need, your right click it and delete it. Done._
![delete-remote-simple-demo](https://user-images.githubusercontent.com/483659/32785823-5818752c-c94a-11e7-9a23-cf63ff4e8888.gif)

The checkbox list is a solution for the edge case or a power user scenario at best and UI should be simple by default (without becoming useless of course). Given that primary scenario for actions on specific commit are targeted to **that** commit, so showing all unrelated branches to a commit is information overload IMHO.

jbialobr commented:
> You see a branch you don't need, your right click it and delete it. Done.

I agree, for me in most cases I want to delete both the local and the remote branch. I usually do it after the branch being merged. Perhaps GUI could offer to do it at once.

mdonatas commented:
@jbialobr handling local/remote deletion would be rather complicated from implementation side of things as well as from the multiple confirmation dialogs, etc.
How about we move on with this minimalistic implementation now and think about further changes if there is user demand?

jbialobr commented:
>handling local/remote deletion would be rather complicated from implementation side of things as well as from the multiple confirmation dialogs, etc.

If everyone agrees that the most common case is to delete a single branch or to delete a single local branch and its remote tracking branch then I suggest to simplify the existing delete branch dialog to have a single branch with no ability to select more branches. (It is very unlikely for me to delete a branch not seeing it on the graph) Then for a local branch we can offer to delete the remote branch along the lines of the Delete tag dialog does. It should support deleting a single remote branch as well.

>How about we move on with this minimalistic implementation now and think about further changes if there is user demand?

I don't get why there is a need to check the "Delete remote branch" checkbox. A remote branch was already chosen from the context menu.

Gua-naiko-che commented:
@jbialobr As it has been discussed in [this related issue](https://github.com/gitextensions/gitextensions/issues/1583), the deletion of a remote branch should be "a little more difficult". Therefore, it was proposed that the dialog for deleting remote branches should different from the one used for deleting local branches. The main difference is precisely the "Delete remote branch" checkbox which acts as an extra confirmation step.

jbialobr commented:
When both local and remote branches are merged in it should be less difficult to delete a remote branch.
The additional confirmation should be required only if we know or we are not sure that the operation is dangerous.
Please, provide at least counterpart settings to disable these additional confirmations.
https://git-extensions-documentation.readthedocs.io/en/latest/settings.html#advanced-confirmations

RussKie commented:
I'm fine with selecting a remote from the context menu and a confirmation dialog without multi selection.
IMO it may not worth the effort combining deletion of local and remote.

I just thought of an edge case for deletion of remote.
If I do squash merge than the remote branch is gone but local git still thinks it exists (there's ref file under .git folder). If you try to delete (i.e. git push) it says that the remote no longer exists.
Should we ensure that the local ref file is deleted?

vbjay commented:
The reason I suggest the remote dropdown is that context menus do cause the
problem of I meant to click on menu option x and I clicked Y because of the
menu still rolling out, or mouse moved as I unclicked...  Also the list of
branches should be filtered by that remote.  If I have origin remote
selected, only origin branches show.

On Tue, Nov 14, 2017 at 4:04 PM RussKie <notifications@github.com> wrote:

> I'm fine with selecting a remote from the context menu and a confirmation
> dialog without multi selection.
> IMO it may not worth the effort combining deletion of local and remote.
>
> I just thought of an edge case for deletion of remote.
> If I do squash merge than the remote branch is gone but local git still
> thinks it exists (there's ref file under .git folder). If you try to delete
> (i.e. git push) it says that the remote no longer exists.
> Should we ensure that the local ref file is deleted?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4146#issuecomment-344397940>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsRB-r0033_zxpy3VE6N_l3EdYPRTks5s2gBlgaJpZM4Qa9Bv>
> .
>

mrexodia commented:
I see this actually deletes the branch from the remote instead of removing the tracking branch (I would expect this to execute `git branch -d -r origin/branch`, not `git push something`). Adding support for deleting remote tracking branches was on my personal list, so perhaps the dialog should be more clear about what it does. For example **this permanently deletes the branch from the remote repository** or something along those lines.

vbjay commented:
There is a plug-in that cleans up tracking branches already.

On Sat, Nov 18, 2017, 8:23 AM Duncan Ogilvie <notifications@github.com>
wrote:

> I see this actually deletes the branch from the remote instead of removing
> the tracking branch (I would expect this to execute git branch -d -r
> origin/branch, not git push something). Adding support for deleting
> remote tracking branches was on my personal list, so perhaps the dialog
> should be more clear about what it does. For example *this permanently
> deletes the branch from the remote repository* or something along those
> lines.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4146#issuecomment-345442078>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsTQD8k-1mz008pGW8WLvndbGyflCks5s3tpdgaJpZM4Qa9Bv>
> .
>

mrexodia commented:
Which plugin do you mean?

Sometimes I temporarily pull a branch from someone else's repository, play around with it and then I want to remove it. Delete branch keeps the remote tracking branch though and I can only seem to remove it from the command line...

XopherThomas commented:
Communication seems to have died here.  Is this pull request going to be resolved and integrated before 2.51?

RussKie commented:
It will be. I am overworked and sleep-deprived at the moment.
There are a number of PRs that all need review.

XopherThomas commented:
Thank you RussKie.  I appreciate your efforts to make GitExtensions even better.

RussKie commented on GitCommands/Git/GitDeleteRemoteBranchCmd.cs:
```diff
@@ -0,0 +1,55 @@
+
+            foreach (var branch in branches)
+            {
+                yield return " :\"" + branch.LocalName + "\"";
```
the leading space isn't necessary

RussKie commented on GitCommands/Git/GitDeleteRemoteBranchCmd.cs:
```diff
@@ -0,0 +1,55 @@
+
+            if (this.branches.Any(b => b.Remote != this.remote))
+            {
+                throw new ArgumentException($"Branch remote missmatch. Branch {this.branches.First(b => b.Remote != this.remote).CompleteName} does not belong to remote {remote}");
```
typo "mismatch"

RussKie commented on GitUI/CommandsDialogs/FormDeleteRemoteBranch.cs:
```diff
@@ -0,0 +1,119 @@
+
+                if (hasUnmergedBranches)
+                {
+                    if (MessageBox.Show(this, _confirmDeleteUnmergedRemoteBranchMessage.Text, _deleteRemoteBranchesCaption.Text, MessageBoxButtons.YesNo) != DialogResult.Yes)
```
I think it only makes sense to show this message if the remote doesn't have any reachable commits on top of it
For example, in this situation the message this warning is not appropriate:

![image](https://user-images.githubusercontent.com/4403806/33739134-0fbe631e-dbf0-11e7-8712-0303f0692ffa.png)

Thoughts?

mdonatas replied:
I am not sure if there is a reliable way to detect that for remote branches. For local ones it's all clear, you have all the data, for remote ones you might not.

RussKie commented:
@mdonatas: we are getting very close to making a cut for 2.51. If we can't finalise the PR in the next few weeks it will miss the boat...

mdonatas commented:
@RussKie I've fixed the typo and am not sure if we can reliably do anything about the unmerged branch warning. Apart from these things, do you expect me to make other changes?
Oh, I've also changed the confirmation checkbox label as per @mrexodia comment to: _Delete branche(s) from remote repository_

I think this is a good solution. I like the workflow of deleting local branches and this mimics it really close. This puts me in a tight spot as there have been other proposals for the workflow/UI which I am not too keen on doing 😇

May I suggest we leave the workflow as it is implemented now in this PR and make further enhancements as per user comments/requests?

mdonatas commented:
Woah @RussKie, that was fast!

Just a fun bit, first working version of this was created ~4 years ago. So glad to see closure after so much time :)

RussKie commented:
You're welcome, thank you for the efforts.

XopherThomas commented:
Thank you all for this feature.  I know I'm not the only requester, but I know at my organization, we will really appreciate the improvement to our workflow.

BUInvent commented:
Thanks a lot for this. I like the new feature :)
